### PR TITLE
[WIP] Merges efforts of modular seurat wrappers (artbio-mblue9-ebi-sanger)

### PR DESCRIPTION
**Pull request by pcm32**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [x] - This PR does something else (explain below)

Starts to unify the seurat wrapping efforts from @mblue9 (https://github.com/galaxyproject/tools-iuc/pull/2047, https://github.com/mblue9/tools-iuc/tree/seurat_scripts/tools/seurat_scripts) @bellenger (https://github.com/galaxyproject/tools-iuc/pull/2195) and EBI/Sanger (https://github.com/ebi-gene-expression-group/container-galaxy-sc-tertiary/commits/develop/tools/tertiary-analysis/seurat). This branch includes all relevant commits cherry picked (plus some copies from our repo, which has a different structure).

FOR REVIEWER:
* [ ] .shed.yml file ok
    - [ ] Toolshed user `iuc` has access to associated toolshed repo(s)
* [ ] Indentation is correct (4 spaces)
* [ ] Tool version/build ok
* [ ] `<command/>`
  - [ ] Text parameters, input and output files `'single quoted'`
  - [ ] Use of `<![CDATA[ ... ]]>` tags
  - [ ] Parameters of type `text` or having `optional="true"` attribute are checked with `if str($param)` before being used
* [ ] Data parameters have a `format` attribute containing datatypes recognised by Galaxy
* [ ] Tests
  - [ ] Parameters are reasonably covered
  - [ ] Test files are appropriate
* [ ] Help
  - [ ] Valid restructuredText and uses `<![CDATA[ ... ]]>` tags
* [ ] Complies with other best practice in [Best Practices Doc](http://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html)

### CODE REVIEW

**Comment by pcm32** [comment]
I intend now to revise the main two alternatives we have on each of the methods. Unless that there is opposition (for instance, someone deciding to provide a monolithic seurat wrapper), I would suggest that we leave all Galaxy wrappers on the `seurat` folder.

# Clustering method

Underlying R functionality (embedded, corresponding to the artbio approach [seurat-clustering](https://github.com/galaxyproject/tools-iuc/blob/adc81f511de57d58e7810265aef116de61d000db/tools/seurat/seurat-clustering.R), and [seurat-find-clusters.R](https://github.com/ebi-gene-expression-group/r-seurat-scripts/blob/master/seurat-find-clusters.R)). 

| Functionality               | Embedded | r-seurat-scripts |
|-----------------------------|----------|------------------|
| Resolution                  | X        | X                |
| Number of components        | X        | X                |
| Which components            |          | X                |
| Reduction type              |          | X                |
| Dimensions                  |          | X                |
| k (for nearest neighbour)   |          | X                |
| SNN Options                 |          | X                |
| Mod. optimization algorithm |          | X                |
| PDF of clusters generation | X         |                  |
| tmp file location |                | X                |

In this case I would suggest that we use the r-seurat-scripts as it is more feature rich. We purposely avoided adding plots to the main steps, as users could do this with the objects on a separate module if desired, reducing load for automated processing. However, we could add the PDF plot generation if needed as an option (although seurat_scripts/seurat_dim_plot.xml provides this modularly).

At the level of the Galaxy wrapper, the artbio wrapper seurat/seurat-clustering is richer than the seurat-scripts/seurat_find_clusters wrapper, so I would suggest to use the first one, adapting it to work with the r-seurat-script layer.

# Filtering

Underlying R functionality:

| Functionality                   | Embedded | r-seurat-scripts |
|---------------------------------|----------|------------------|
| Cell subset names               |          | X                |
| Cell low threshold              | X        | X                |
| Cell high threshold             | X        | X                |
| Specific cells to use           |          | X                |
| Min cells (for object creation) | X        |                  |
| Min genes (for object creation) | X        |                  |
| Reads serialized R object       |          | X                |
| Reads counts file               | X        |                  |
| VlnPlot PDF                     | X        |                  |
| GenePlot PDF                    | X        |                  |
| Choose parameter for filtering  |          | X                |

...to be continued

**Comment by bellenger-l** [comment]
Hello everyone,

Is it possible to get reviewers for this PR, so that we can move on with these tools ?

Thanks,
Lea

**Comment by pcm32** [comment]
Hi, not done on this yet (WIP), was hoping to get back to it soon. But happy to get some help. I probably have some commits to push.

**Comment by pcm32** [comment]
It would help as well if someone keeps doing the exercise above (of functionality comparison). I'm currently a bit stuck with the part of adding the additionally needed functionality to r-seurat-scripts to have filtering behaving as done by  @bellenger-l .

**Comment by drosofff** [comment]
> It would help as well if someone keeps doing the exercise above (of functionality comparison).

Sorry @pcm32 but I am a bit lost: what can we do to help, specifically ?

**Comment by pcm32** [comment]
@drosofff I meant this exercise: https://github.com/galaxyproject/tools-iuc/pull/2231#issuecomment-452707630

That is, analyse the different functionality bit by bit and see how we merge it. Also comment on the current ones and express agreement/disagreement. I normally wrote one of those, then waited for a few days, and under no opposition I implemented. For us our only requirement is to work with the intermediate scripting layer so that we can easily use this from other workflow setups.

**Comment by bgruening** [comment]
I will make some time this weekend and look over it. Sorry for the delay.

File: tools/seurat/seurat-clustering.xml [hunk 1]
```diff
@@ -0,0 +1,103 @@
+<tool id="seurat-clustering" name="Seurat clustering" version="2.3.1+galaxy0">
+    <description>find clusters of cells</description>
+    <macros>
+        <import>seurat_macros.xml</import>
+    </macros>
+    <expand macro="requirements" />
+    <expand macro="version" />
+    <command detect_errors="exit_code"><![CDATA[
+#if $rscript:
+  cp \$(which seurat-find-clusters.R) '$out_rscript' &&
```
[the remark this answers is not in the document]
Comment (pcm32) [reply tools/seurat/seurat-clustering.xml]: I agree

File: tools/seurat/seurat-clustering.xml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+--input-object-file '$data_rds'
+--output-object-file '$out_rds'
+--output-text-file output_tab
+#if $adv.resolution:
+--resolution '$adv.resolution'
+#end if
+
+#if $adv.numPCs:
+--dims-use \$(seq -s , 1 '$adv.numPCs')
```
[the remark this answers is not in the document]
Comment (pcm32) [reply tools/seurat/seurat-clustering.xml]: I guess we can put a validator for integer <= 20 or a reasonable value.

File: tools/seurat/seurat-clustering.xml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+#end if
+
+## TODO add pdf support as optional
+]]></command>
+
+    <inputs>
+        <param name="data_rds" type="data" format="rdata" label="Seurat object" help="This should be a RDS file of a single Seurat object."/>
+        <param name="rscript" type="boolean" truevalue="TRUE" falsevalue="FALSE" checked="False" label="Output Rscript?" help="If this option is set to Yes, the Rscript used by this tool will be provided as a text file in the output. Default: No" />
```
[the remark this answers is not in the document]
Comment (pcm32) [reply tools/seurat/seurat-clustering.xml]: I intend to remove all of this. This wasn't really ready for review, it was an initial merge :-).

**Comment by pcm32** [comment]
I can comment that I have updated some of our Seurat modules for other purposes, will add them here as soon as possible. It should make merging easier I think and they expose nearly all the functionality of the underlying scripts package.

**Comment by pcm32** [comment]
I would like to make people aware of https://github.com/ebi-gene-expression-group/container-galaxy-sc-tertiary/pull/57 which has our latest versions for all the seurat tools. Any further comparisons should be made against that. I will do my best to dump all those changes here as well soon. It is a good improvement of documentation and it exposes nearly all functionality available in seurat-scripts.

**Comment by bgruening** [comment]
Keep it coming @pcm32, that is awesome!

File: tools/seurat/seurat_macros.xml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0"?>
+<macros>
+
+    <token name="@VERSION@">0.0.4</token>
+
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="@VERSION@">r-seurat-scripts</requirement>
```
[the remark this answers is not in the document]
Comment (pcm32) [reply tools/seurat/seurat_macros.xml]: We could do that... You mean something like <seurat-version>_<scripts-version>?

[the remark this answers is not in the document]
Comment (bgruening) [reply tools/seurat/seurat_macros.xml]: Yes, something like that.

**Comment by drosofff** [comment]
Hi @pcm32, could you comment on the purpose on this WIP given

- https://toolshed.g2.bx.psu.edu/view/ebi-gxa/seurat_create_seurat_object/f8865ae30723
- https://toolshed.g2.bx.psu.edu/view/ebi-gxa/seurat_dim_plot/6e26ee99fdfc
- https://toolshed.g2.bx.psu.edu/view/ebi-gxa/seurat_export_cellbrowser/8c2b1d2659df
- https://toolshed.g2.bx.psu.edu/view/ebi-gxa/seurat_filter_cells/c1848daebad1
- and 8 other repo all uploaded to the toolshed in April, 3th

?

I understand that directly contributing to https://github.com/ebi-gene-expression-group/container-galaxy-sc-tertiary.git would avoid further duplication effort.

Thanks for saving our time.

**Comment by pcm32** [comment]
Hi @drosofff yes, unfortunately I could never find time to continue this merge and had to push the setup we had to the tool shed to use it from our multiple production instances. So, yes, you could contribute at the other repo you mention and we could eventually move it to tools IUC when it is good enough (I needed to have told on the tool shed but process here takes too long). Apologies for any inconvenience or lack of communication. I will choose this then.

**Pull request closed without merging**